src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from contextlib import contextmanager
+
+from cryptography.utils import cached_property
+from snowflake.cli._plugins.nativeapp.sf_facade_exceptions import (
+    CouldNotUseObjectError,
+    UserScriptError,
+    handle_unclassified_error,
+)
+from snowflake.cli.api.constants import ObjectType
+from snowflake.cli.api.entities.common import get_sql_executor
+from snowflake.cli.api.errno import (
+    DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED,
+    NO_WAREHOUSE_SELECTED_IN_SESSION,
+)
+from snowflake.cli.api.project.util import to_identifier
+from snowflake.cli.api.sql_execution import SqlExecutor
+from snowflake.connector import ProgrammingError
+
+
+class SnowflakeSQLFacade:
+    def __init__(self, sql_executor: SqlExecutor | None = None):
+        self._sql_executor = (
+            sql_executor if sql_executor is not None else get_sql_executor()
+        )
+
+    @cached_property
+    def _log(self):
+        return logging.getLogger(__name__)
+
+    def _use_object(self, object_type: ObjectType, name: str):
+        """
+        Call sql to use snowflake object with error handling
+        @param object_type: ObjectType, type of snowflake object to use
+        @param name: object name, has to be a valid snowflake identifier.
+        """
+        try:
+            self._sql_executor.execute_query(f"use {object_type} {name}")
+        except ProgrammingError as err:
+            if err.errno == DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED:
+                raise CouldNotUseObjectError(object_type, name) from err
+            else:
+                handle_unclassified_error(err, f"Failed to use {object_type} {name}.")
+        except Exception as err:
+            handle_unclassified_error(err, f"Failed to use {object_type} {name}.")
+
+    @contextmanager
+    def _use_warehouse_optional(self, new_wh: str | None):
```

Comment:
Was wondering what the design philosophy was making all of these optional? It seems like the original ones require non-optional arguments, it feels a bit weird to me to be able to call `use_warehouse_optional(None)` and have that be a noop, from an API standpoint; maybe it would be better to be explicit about our invocations?

Reply:
Notice these are internal. The general idea is that we're expressing high-level operations in the public facade API, and we let the caller specify the parameters. The only API we're introducing is for user scripts. It's a general API, so using the session role is a valid option, for example. Our NA use does prefer using the package role if one was specified. However, the abstractions here would let us move the fallback "use role" to the facade for cases where the role is optional. The helpers were written this way so that we can make the rest of the implementation clean. This is really just a "temporarily activate this WH/role/db if one was provided" helper.